src/encode.cpp

```diff
@@ -936,6 +1076,43 @@ void encodeOp(State &st, mlir::arith::SelectOp op, bool) {
   }
 }
 
+template<>
+void encodeOp(State &st, mlir::func::CallOp op, bool) {
+  if (op.getNumResults() != 1) {
+    throw UnsupportedException(
+      op.getOperation(),
+      "Invalid number of return values");
+  }
+  
+  const auto callee = op.getCallee(); 
+  if (!getFunction(callee)) {
+    vector<mlir::Type> domain(op.getOperandTypes().begin(),
+                              op.getOperandTypes().end());
+    auto range = op.getResultTypes().front();
+    try {
+      declareFunction(move(domain), move(range), move(callee));
+    } catch (UnsupportedException e) {
+      throw UnsupportedException(op.getOperation(), e.getReason());
+    }
+  }
+
+  vector<ValueTy> operands;
+  operands.reserve(op.getNumOperands());
+  for (const auto& operand: op.getOperands()) {
+    const auto operandTy = operand.getType();
+    if (operandTy.isa<mlir::FloatType>()) {
+      operands.push_back(st.regs.get<Float>(operand));
+    } else if (operandTy.isa<mlir::IntegerType>()) {
+      operands.push_back(st.regs.get<Integer>(operand));
+    } else if (operandTy.isa<mlir::IndexType>()) {
+      operands.push_back(st.regs.get<Index>(operand));
+    }
+  }
+
+  auto calleeUF = *getFunction(callee);
+  st.regs.add(op.getResult(0), calleeUF.apply(operands));
```

Comment:
Let's emit a warning message here saying that we assume that call ops are stateless (does not read or write global memory).

Reply:
Added

Reply:
I think we're getting a bit too many warnings (at least 3 times per validation pair). Is this okay, or do you have some better place for the warning?

Reply:
Moved the message to `declareFunction`